// @coreui/angular/esm2015/lib/sidebar/app-sidebar-minimizer.component.js
import { Component, HostBinding, HostListener, Input } from '@angular/core';
import { AppSidebarService } from './app-sidebar.service';
export class AppSidebarMinimizerComponent {
    constructor(sidebarService) {
        this.sidebarService = sidebarService;
        this.role = 'button';
        this.sidebarMinimizerClass = true;
    }
    toggleOpen($event) {
        $event.preventDefault();
        this.sidebarService.toggle({ minimize: 'toggle' });
    }
}
AppSidebarMinimizerComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-sidebar-minimizer, cui-sidebar-minimizer',
                template: ``
            },] }
];
AppSidebarMinimizerComponent.ctorParameters = () => [
    { type: AppSidebarService }
];
AppSidebarMinimizerComponent.propDecorators = {
    role: [{ type: HostBinding, args: ['attr.role',] }, { type: Input }],
    sidebarMinimizerClass: [{ type: HostBinding, args: ['class.sidebar-minimizer',] }],
    toggleOpen: [{ type: HostListener, args: ['click', ['$event'],] }]
};
