// @coreui/angular/esm2015/lib/sidebar/app-sidebar-nav.component.js
import { Component, HostBinding, Input } from '@angular/core';
import { Router } from '@angular/router';
export class AppSidebarNavComponent {
    constructor(router) {
        this.router = router;
        this.navItems = [];
        this.sidebarNavClass = true;
        this.role = 'nav';
        this.navItemsArray = [];
    }
    ngOnChanges(changes) {
        this.navItemsArray = Array.isArray(this.navItems) ? this.navItems.slice() : [];
    }
}
AppSidebarNavComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-sidebar-nav, cui-sidebar-nav',
                template: "<app-sidebar-nav-items\n  class=\"nav\"\n  [items]=\"navItemsArray\">\n</app-sidebar-nav-items>\n"
            },] }
];
AppSidebarNavComponent.ctorParameters = () => [
    { type: Router }
];
AppSidebarNavComponent.propDecorators = {
    navItems: [{ type: Input }],
    sidebarNavClass: [{ type: HostBinding, args: ['class.sidebar-nav',] }],
    role: [{ type: HostBinding, args: ['attr.role',] }, { type: Input }]
};
